File: algorithmclass/wordcreator.cpp
#include <algorithm>
#include <unordered_set>
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>

using namespace std;

string
get_string(bool memo, int *location, string instring)
{
    string outstring = "";
    if(!memo){ 
        outstring = "No, cannot be split\n";  
    }else{
        outstring = "Yes, can be split\n";
        for(int i = 0; i < instring.size() - 1; i++){
            outstring += instring.substr(i,location[i]+1) + " ";
            i          = location[i];
        }
    }
    return outstring;
}

/*bool
recursive_find(unordered_set<string>&dictionary, bool *memo, int *location, string instring, int i=0)
{
    if(i == instring.size())
        return true;
    
    if(memo[i])
        return memo[i];
    
    memo[i] = false;
    for(int j = i; i < instring.size(); j++)
        if((dictionary.find(instring.substr(i,j+1)) != dictionary.end()) && recursive_find(dictionary,memo,location,instring,j+1)){
            memo[i]     = true;
            location[i] = j;
        }
    return memo[i];
}*/

string
iterative_find(unordered_set<string>&dictionary, string instring)
{
    bool   memo[instring.size()+1];
    int    location[instring.size()+1];
    string outstring("");
    string substring("");

    memo[instring.size()] = true;
    for(int i = instring.size()-1; i >= 0; i--){
        memo[i] = false;
        for(int j = i; j < instring.size(); j++)
            if((dictionary.find(instring.substr(i,j+1)) != dictionary.end())  && memo[j+1]){
                memo[i]     = true;
                location[i] = j; 
            }
    }

    return get_string(memo[0],location,instring);
}

void
initialize_array(bool* memo, int *location, int size)
{
    for(int i = 0; i < size; i++){
        memo[i]     = false;
        location[i] = 0; 
    }
}

int
main()
{
    unordered_set<string> dictionary(58114);
    ifstream myfile("diction10k.txt");

    int           numphrase(0);
    bool          found(false);
    string        in("");

    //init dictionary
    while(getline(myfile, in)){
        istringstream iss(in);
        iss >> in;
        dictionary.emplace(in);
    } 

    cin >> numphrase; 
    string phrases[numphrase];
    for(int i = 0; i < numphrase; i++)
        cin >> phrases[i];
    
    for(int i = 0; i < numphrase; i++){
        cout << "phrase number: 1" << endl;
        cout << phrases[i] << endl << endl;
        cout << "iterative attempt:" << endl;
        cout << iterative_find(dictionary,phrases[i]) << endl; 
        cout << "memoized attempt:" << endl;
//        int  location[phrases[i].size()+1];
//        bool memo[phrases[i].size()+1];
//        initialize_array(memo,location,phrases[i].size()+1);
//        found = recursive_find(dictionary,memo,location,phrases[i]);
//        for(int i = 0; i < phrases[i].size();i++)
//            cout << location[i];
//        cout << get_string(found,location,phrases[i]) << endl << endl;
        
    }

    myfile.close();
    
    return 1;
}

File: algorithmclass/wordcreator.py
def init_dictionary():
    hash_set = set();
    file = open('diction10k.txt','r')
    for line in file:
        hash_set.add(line.strip('\n')) 
    
    file.close()
    return hash_set

dictionary = init_dictionary()
memo       = list()
location   = list()

def recursive_find(string,i=0):
    global memo
    global location
    global dictionary

    if i == len(string):
        return True
    
    if memo[i] != -1:
        return memo[i]
    
    memo[i] = False
    j = i
    while(j < len(string)):
        found = string[i:j+1] in dictionary
        if(found and recursive_find(string,j+1)):
            memo[i]     = True
            location[i] = j
        j += 1 
    
    return memo[i] 
         
def iterative_find(string):
    global memo
    global location
    global dictionary

    memo     = [False for _ in range(len(string)+1)] 
    location = [0 for _ in range(len(string)+1)] 
    
    
    memo[len(string)] = True
    i = len(string)-1
    while(i >= 0):
        memo[i] = False
        j = i; 
        while(j < len(string)):
            found = string[i:j+1] in dictionary
            if(found and memo[j+1]):
                memo[i]     = True;
                location[i] = j;
            j+=1
        i = i - 1

    print_string(string, memo[0])


def print_string(string, can_be_split):
    global location

    if(not can_be_split):
        print("No, cannot be split\n")
        return
    
    outstring = "Yes, can be split\n"
    i = 0
    while(i < len(string)-1):
        outstring += string[i:location[i]+1] + " "
        i          = location[i]
        i         += 1
   
    print(outstring + "\n")

 
def main():
    global memo
    global location
    global dictionary

    amount_phrase = int(input())
    phrases = []
    for _ in range(amount_phrase):
        phrases.append(input())  
    
    for i in range(amount_phrase):
        print("phrase number: " + str(i+1))
        print(phrases[i],end="\n\n")
        print("iterative attempt:")
        iterative_find(phrases[i])
        memo     = [-1 for _ in range(len(phrases[i])+1)] 
        location = [0 for _ in range(len(phrases[i])+1)] 
        print("memoized attempt:")
        found = recursive_find(phrases[i])
        print_string(phrases[i],found)
        print()

main()
